src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt

```diff
@@ -469,22 +552,61 @@ abstract class AbstractCheckTask(
     private fun Observable<StreamContainer>.continueObserve(sessionKey: SessionKey, sequence: Long): Observable<Message> =
         filter { streamContainer -> streamContainer.sessionKey == sessionKey }
             .flatMap(StreamContainer::bufferedMessages)
-            .filter { message -> message.metadata.id.sequence > sequence }
+            .filter { message ->
+                if (message.metadata.id.sequence == sequence) {
+                    bufferContainsStartMessage = true
+                }
+                message.metadata.id.sequence > sequence
+            }
 
-    private fun Checkpoint.getSequence(sessionKey: SessionKey): Long {
-        val sequence = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
-            ?.directionToSequenceMap?.get(sessionKey.direction.number)
+    private fun Checkpoint.getCheckpointData(sessionKey: SessionKey): CheckpointData {
+        val checkpointData = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
+            ?.directionToCheckpointDataMap?.get(sessionKey.direction.number)
 
-        if (sequence == null) {
+        if (checkpointData == null) {
             if (LOGGER.isWarnEnabled) {
-                LOGGER.warn("Checkpoint '{}' doesn't contain sequence for session '{}'", shortDebugString(this), sessionKey)
+                LOGGER.warn("Checkpoint '{}' doesn't contain checkpoint data for session '{}'", this.toJson(), sessionKey)
             }
-        } else {
+            val sequence = sessionAliasToDirectionCheckpointMap[sessionKey.sessionAlias]
+                ?.directionToSequenceMap?.get(sessionKey.direction.number)
+            if (sequence == null) {
+                if (LOGGER.isWarnEnabled) {
+                    LOGGER.warn("Checkpoint '{}' doesn't contain sequence for session '{}'", this.toJson(), sessionKey)
+                }
+                return CheckpointData(DEFAULT_SEQUENCE)
+            }
+            return CheckpointData(sequence)
+        }
+
+        return checkpointData.convert().apply {
             LOGGER.info("Use sequence '{}' from checkpoint for session '{}'", sequence, sessionKey)
         }
+    }
 
-        return sequence ?: DEFAULT_SEQUENCE
+    private fun checkOnMessageTimeout(timestamp: Timestamp): Boolean {
+        return checkpointTimeout == null || checkpointTimeout!!.isAfter(timestamp) || checkpointTimeout == timestamp
```

Comment:
I believe we don't need `!!` because we get there only if the `checkpointTimeout != null`

Reply:
Kotlin cannot smart cast 'Timestamp?' to 'Timestamp' here